src/field/generated_fieldops.cxx

```diff
@@ -0,0 +1,1668 @@
+
+#if CHECK > 0
+  if (lhs.getLocation() != rhs.getLocation()) {
+    throw BoutException("Error in subtraction: fields at different locations. lhs is at "
+                        "%s, rhs is at %s!",
+                        strLocation(lhs.getLocation()), strLocation(rhs.getLocation()));
+  }
+#endif
+
+  result.setLocation(rhs.getLocation());
+
+  checkData(result);
+  return result;
+}
+
+// Provide the C function to update Field3D by subtraction with Field3D
+void autogen_Field3D_Field3D_subtraction(BoutReal *__restrict__ lhs,
+                                         const BoutReal *__restrict__ rhs, int len) {
+
+  for (int i = 0; i < len; ++i) {
+    lhs[i] -= rhs[i];
+  }
+}
+
+// Provide the C++ operator to update Field3D by subtraction with Field3D
+Field3D &Field3D::operator-=(const Field3D &rhs) {
+  // only if data is unique we update the field
+  // otherwise just call the non-inplace version
+  if (data.unique()) {
+    Indices i{0, 0, 0};
+
+    ASSERT1(fieldmesh == rhs.getMesh());
+
+    checkData(*this);
+    checkData(rhs);
+
+    autogen_Field3D_Field3D_subtraction(&(*this)[i], &rhs[i],
+                                        fieldmesh->LocalNx * fieldmesh->LocalNy *
+                                            fieldmesh->LocalNz);
+
+#if CHECK > 0
+    if (this->getLocation() != rhs.getLocation()) {
+      throw BoutException("Error in subtraction: fields at different locations. lhs is "
+                          "at %s, rhs is at %s!",
+                          strLocation(this->getLocation()),
+                          strLocation(rhs.getLocation()));
+    }
+#endif
+    checkData(*this);
+
+  } else {
+    (*this) = (*this) - rhs;
+  }
+  return *this;
+}
+
+// Do the actual multiplication of Field3D and Field2D
+void autogen_Field3D_Field3D_Field2D_multiplication(BoutReal *__restrict__ result,
+                                                    const BoutReal *__restrict__ lhs,
+                                                    const BoutReal *__restrict__ rhs,
+                                                    int nx, int ny, int nz) {
+
+  for (int x = 0; x < nx; ++x)
+    for (int y = 0; y < ny; ++y)
+      for (int z = 0; z < nz; ++z) {
+        result[z + nz * (y + ny * x)] = lhs[z + nz * (y + ny * x)] * rhs[y + x * ny];
+      }
+}
+
+// Provide the C++ wrapper for multiplication of Field3D and Field2D
+Field3D operator*(const Field3D &lhs, const Field2D &rhs) {
+  Indices i{0, 0, 0};
+  Mesh *localmesh = lhs.getMesh();
+
+  ASSERT1(localmesh == rhs.getMesh());
+
+  Field3D result(localmesh);
+  result.allocate();
+  checkData(lhs);
+  checkData(rhs);
+
+  autogen_Field3D_Field3D_Field2D_multiplication(&result[i], &lhs[i], &rhs[i],
+                                                 localmesh->LocalNx, localmesh->LocalNy,
+                                                 localmesh->LocalNz);
+
+  result.setLocation(lhs.getLocation());
+
```

Comment:
No location checks? Whilst Field2D doesn't have a location do we need to ensure the Field3D is therefore at the cell centre?

Reply:
No. Otherwise the staggered grids are unusable.

See e.g. #64 - that would require us to extend the staggered grids support to field2D - but that is non-trivial.

Reply:
OK, but does this mean they are usable but not necessarily correct?

Reply:
To be honest, I am using it for a slab geometry, so dx etc are all constant, and the non-staggered-ness of Field2D does not bother me.

I don't know whether anybody quantified the issues for other meshes, that have changes along the staggered direction.

To solve this in a nice way would probably require to wrap the mesh quantities in an object, that contains them at several locations, and depending on the field's location they are operated with, to choose the appropriate field. A potentially remaining issue is, if one of the fields is changed, to update the other ones.

The other approach would be having a different mesh for the staggered quantities - but that results in challanges with taking derivatives from one mesh, on to the the other mesh, and vice versa, i.e. we need to add relations between meshes, both in terms of interpolating, and differentiating.

Reply:
Thanks for the detailed response and some interesting ideas for how to potentially improve this. 

I'm slightly concerned that if users try to use staggered fields in their physics model and are also using Field2D for some things they may not realise the situation may not be as they think. It's now obvious when using two 3D field at different locations that this is bad (at least when CHECK is >0) but this may convince some users that their use of staggered Field3D and a non-staggered Field2D is fine as it doesn't throw an exception even with all checks turned on. I'm not suggesting there's an easy fix for this (as you've noted) and breaking staggered grids for everyone isn't a good idea either -- is there anything we can do to alert/warn the user to this beyond the manual?

Reply:
I think this conversation is off topic.
I does not get worse with this PR, it is just not jet perfect.
I will open an issue to continue the discussion there.